packages/lesswrong/server/resolvers/ultraFeedResolver.ts

```diff
@@ -169,118 +138,97 @@ const createSourcesMap = (
-    if (value.items.length > 0) {
-      acc[key as FeedItemSourceType] = value;
-    }
-    return acc;
-  }, {} as Record<FeedItemSourceType, WeightedSource>);
+    bucket.items.push(itm);
+  };
+
+  spotlightItems.forEach(s => push("spotlights", { type: "feedSpotlight", feedSpotlight: s }));
+
+  postThreadsItems.forEach(p =>
+    (p.postMetaInfo?.sources ?? []).forEach(src =>
+      push(src as FeedItemSourceType, { type: "feedPostWithContents", feedPost: p })));
+
+  commentThreadsItems.forEach(t =>
+    push("recentComments", { type: "feedCommentThread", feedCommentThread: t }));
+
+  bookmarkItems.forEach(item => push("bookmarks", item));
+
+  return Object.fromEntries(Object.entries(sources).filter(([, v]) => v.items.length));
 };
 
 /**
  * Extract IDs that need to be loaded from sampled items
  */
-const extractIdsToLoad = (sampledItems: SampledItem[]) => {
-  const spotlightIdsToLoad: string[] = [];
-  const commentIdsToLoad = new Set<string>();
-
-  sampledItems.forEach(item => {
-    if (item.renderAsType === "feedSpotlight") {
-      spotlightIdsToLoad.push(item.feedSpotlight.spotlightId);
-    } else if (item.renderAsType === "feedCommentThread") {
-      item.feedCommentThread.comments?.forEach(comment => {
-        if (comment.commentId) {
-          commentIdsToLoad.add(comment.commentId);
-        }
-      });
+const extractIdsToLoad = (sampled: SampledItem[]) => {
+  const postIds: string[] = [];
+  const spotlightIds: string[] = [];
+  const commentIdsSet = new Set<string>();
+
+  sampled.forEach(it => {
+    if (it.type === "feedSpotlight") {
+      spotlightIds.push(it.feedSpotlight.spotlightId);
+    } else if (it.type === "feedCommentThread") {
+      it.feedCommentThread.comments?.forEach(c => c.commentId && commentIdsSet.add(c.commentId));
+    } else if (it.type === "feedPost") {
+      postIds.push(it.feedPostStub.postId);
     }
   });
-  
+
   return {
-    spotlightIds: spotlightIdsToLoad,
-    commentIds: Array.from(commentIdsToLoad)
+    spotlightIds,
+    commentIds: Array.from(commentIdsSet),
+    postIds
   };
 };
 
 /**
  * Transform sampled items into UltraFeedResolverType results
  */
 const transformItemsForResolver = (
-  sampledItems: SampledItem[],
+  sampled: SampledItem[],
   spotlightsById: Map<string, DbSpotlight>,
-  commentsById: Map<string, DbComment>
+  commentsById: Map<string, DbComment>,
+  postsById: Map<string, DbPost>
 ): UltraFeedResolverType[] => {
-  return filterNonnull(sampledItems.map((item: SampledItem, index: number): UltraFeedResolverType | null => {
-    if (item.renderAsType === "feedSpotlight") {
+  return filterNonnull(sampled.map((item, index): UltraFeedResolverType | null => {
+    if (item.type === "feedSpotlight") {
       const spotlight = spotlightsById.get(item.feedSpotlight.spotlightId);
       if (!spotlight) return null;
       
       return {
-        type: item.renderAsType,
+        type: item.type,
```

Comment:
fyi this is a backwards-breaking change which might be fine now but would be bad if this were widely deployed to users.